### ISSUE TITLE
[RFC] 2FA/SSO considerations for Mailu

### ISSUE THREAD

**Issue by nextgens** [Mailu/Mailu#2222]
In #783 we talk about the building blocks required for 2FA to make any sense; Here I would like to explore options.

## what would we require 2FA for?
- All logins?
- Only some privileged actions? If so, which and do we need more than one level of vertical segregation?
- When we think something has changed? (once in a while/freshness or when we fingerprint a new device)
- When we think something risky is attempted? (not sure how that would apply in the context of Mailu)

# how would provisioning work?
Admin creates a user: does it mean he enrolls the 2FA rigmarole or send the user a temporary password/token that puts him on the enrollment route (and we accept that until it's initialized there is no 2FA)? Is it worth having a more complicated scheme where the administrator has two pieces of info he needs to communicate (potentially on different channels) to the user?

# what would be the fallback/recovery mechanisms?
- Are we happy with just "talk to admin"?

# zero-touch/passwordless auth (where you don't even have to remember your username/email: you show up on the login page, activate your token and you're authenticated)
If we decide to support something of the sort we need to think about it from the get-go.  Usually login processes are "journeys" spanning several steps/pages.

Once the above has been answered we can talk about technologies and user experience:
- [] webauthn (trendy, phish-proof, easy, clearly where the industry is going ATM)
- [] TOTP (rfc6238, standard)
- [] oauth2 client (to enable "login with Google/Facebook/...": that requires configuration)
- [] "send me a link on my backup email"
- [] "send me a text" (would require configuration and $$)
- [] consensus based approval for account resets (get X authenticated users to click on XXX to receive your reset token)

**Comment by Diman0** [comment]
I did some intial brainstorming with nextgens. So far we decided the following:

## when is it required?
- For email client login
- For all web logins
    - but have an exemption after successfull login.
    - fingerprinting/cookie for suppressing 2fa question? we want it per device.
    - Apply this to all sort of logins (webmail, admin). No difference between webmail/admin (KISS!!!)


-> KISS: all logins

## zero-touch/passwordless auth (where you don't even have to remember your username/email: you show up on the login page, activate your token and you're authenticated)
If we decide to support something of the sort we need to think about it from the get-go.  Usually login processes are "journeys" spanning several steps/pages.

-> not for now

## reach | what would we require 2FA for?
- should we disable non-2FA ways of authenticating? (disable IMAP/PLAIN in favour of xoauth2? what about existing tokens?)

-> For everything. Only exception are application token.

- should we allow to "add" alternative factors of authenticating rather than enforce a single factor everywhere?
-> yes. E.g. if you have topt and webathn. Then you can choose which one to use for your login.

- should there be mandatory factors?

-> yes (KISS) / no SSO yet nor zero-touch logins 

## how would provisioning work?
- client (auto)configuration?
    - investigate if existing library can be used such as automx2.
    - Consideration is that we must still be able to use front.
    - For autoconfiguration it is required to have standard domains such as autodiscovery.domain.com
        - we also need certificates for these fixed autodiscovery related sub domains
    - Or we implement ourselves
        - creating the XML is fairly easy with Flask.
- Admin creates a user: does it mean he enrolls the 2FA rigmarole or send the user a temporary password/token that puts him on the enrollment route (and we accept that until it's initialized there is no 2FA)? Is it worth having a more complicated scheme where the administrator has two pieces of info he needs to communicate (potentially on different channels) to the user?
-> start simple / KISS
On login you get the enrollment process.

## what would be the fallback/recovery mechanisms?
- Are we happy with just "talk to admin"?
-> KISS; no recovery. Admin can run mailu flask command to handle losing his second factor
- all users/per domain/per site(mailu.env)?
-> per user is fine but the admin decides; enabled by default for new accounts (unless configured otherwise in  mailu.env. We will only document it. Not part of mailu.env file by default).
- minimum number of technologies/tokens during enrollment?
-> No. Just show a warning if there's only one, let the admin configure which technologies are allowable

## technologies
Once the above has been answered we can talk about technologies and user experience:
- [x] webauthn (trendy, phish-proof, easy, clearly where the industry is going ATM)
- [x] TOTP (rfc6238, standard)
- [later] fingerprinting devices / threat intel / AI
- [later] oauth2 client (to enable "login with Google/Facebook/...": that requires configuration) see #1984
- [later / not by me / not desired] recovery codes
- [maybe - security concerns] "send me a link on my backup email"
- [never] "send me a text" (would require configuration and $$)
- [later / nice to have] consensus based approval for account resets (get X authenticated users to click on XXX to receive your reset token)

**Comment by jessebot** [comment]
I'm not a member of the project, but as someone who frequently needs to make sure they meet certain security standards for auditors, I need to have 2FA enabled on all web interfaces. Adding integration with something like keycloak would be awesome, but even just allowing TOTP would greatly improve adoption.

> what would be the fallback/recovery mechanisms?

On this specific question, I think having some backup recovery codes that are issued at time of 2FA setup would be best, as this is industry standard right now. For now, to get the feature out the door, I agree talking to your admin makes the most sense, but down the line, issuing backup recovery codes seems like a good longer term milestone.

Is there anything I can help with here? I saw https://github.com/Mailu/Mailu/pull/2379#issuecomment-1163666247, but the [link](https://github.com/heviat/Mailu-OpenID/tree/openid-testing) is dead. If there's any testing I can do, or code in python or bash that needs writing, I can take a look. I don't want to step on anyone's toes or duplicate any effort, but I think mailu is a cool project and I'd love to help :)

**Comment by nextgens** [comment]
Since then, we have implemented https://mailu.io/2.0/configuration.html#header-authentication-using-an-external-proxy which enables tools like keycloack to be deployed alongside Mailu without any code change.

As far as I am concerned, I still see value in having both TOTP and webauthn/passkeys supported natively; I have already some code for it and just need to clean it up...

Regarding xoauth2 support on the "mail" side of things, we are kind of stuck: again, the code is mostly (I have a PoC already) ready but none of the MUAs (except mutt) allow for it to be configured... For instance in Thunderbird: https://github.com/thundernest/autoconfig/issues/18 https://bugzilla.mozilla.org/show_bug.cgi?id=1602166

Anything you could do to get MUAs to support [rfc7628](https://datatracker.ietf.org/doc/rfc7628/) or [rfc7591](https://datatracker.ietf.org/doc/rfc7591/) would help :)

**Comment by jcz1** [comment]
The solution with external idP has one major disadvantage. It breaks the password management in the Mailu interface and makes it almost impossible for domain administrators to manage users in Mailu (administration must be done in multiple places and domain administrators must have admin rights in idP).
AFAIK 2FA in webmail does not address 2FA in admin interface. Or how can be Roundcube plugin integrated with Mailu login form?
IMHO it makes sense to use full-featured mail clients with tokens and 2FA for occasional administration/webmail use. And it is much easier to implement 2FA for web clients. Btw i prefer webauthn but OTP has more people.

**Comment by alexsalex** [comment]
> https://mailu.io/2.0/configuration.html#header-authentication-using-an-external-proxy

Doesn't work with Roundcube installed on different standalone host...

**Comment by alexsalex** [comment]
Please, it should be easy implementation.... but HUGE step forward.

**Comment by Gurkengewuerz** [comment]
any new thoughts on this?
three years later a lot of companies implemented 2FA for their endusers. i already can understand the downside of using SSO for mails. a lot of clients can't use the SSO. thats why my opinion is to use a local authentification for mailu.
but addind 2FA and if enabled require app passwords for IMAP/SMTP logins, on the frontend logging in with your password and the 2FA should be easy to do.

**Comment by nextgens** [comment]
Most of what has been discussed has been implemented, some of it has not been pushed upstream though.

-> autoconfig: we have a working implementation; xoauth2 compat is pending on https://datatracker.ietf.org/doc/draft-eggert-mailmaint-uaautoconf/ and/or https://github.com/Mailu/Mailu/issues/2222#issuecomment-1621231333 ; none of it will be useful until MUAs implement it though.
-> TOTP and passkeys: I have each working on a separate branch... I use them but works remain on merging them and upstreaming it.

I had a couple bad interactions with users recently, as as result I am scaling back on the time I volunteer to Mailu. Us not managing to meet is also weighting in on my motivation... so there is no ETA. As you can see no one else has picked it up.

You are talking about companies: those have a tool at their disposal to influence priorities and motivate people: money. Do you know of any that wants to part with some to move this forward?

**Comment by sdomi** [comment]
@nextgens I'm considering pouring some time into this, considering I already... "acclimatized myself" with the relevant parts of the codebase through that MR review yesterday. Also (just so we're clear on this), I don't touch GenAI on moral grounds (as one could guess by clicking on my profile/webpage).

Any pointers / specific areas that still need work? If you have unmerged code, I'd rather take that and improve it than start from scratch myself. My goal is to have something to deploy in 3 months max, but if it takes longer to upstream it, so be it.

**Comment by nextgens** [comment]
I have written most of this already (and been using it in prod), I just need to find/make the time to send a PR.

Will try to prioritise it

**Comment by sdomi** [comment]
ACK! thanks lots
